### PR TITLE
[PyTorch]Fused ScaledActivation Kernel in TE Ops and reducing CPU overheads for MOE

### LINKED ISSUE #2988: Activation + Group Quantize Fusion with te.Sequential
**Is your feature request related to a problem? Please describe.**
Grouped MLP Block in MOE can be roughly -equated to  GroupedLinear + Activation + GroupedLinear. Each of the GroupedLinear needs to go through group_quantize. For the cases where this is unfused we need to atleast enable fusing Activation with the group_quantize happening in the second GroupedLinear layer.

**Describe the solution you'd like**
- Expose the right tex functions in pytorch that fuses activation and quantize. This is already supported for normal pytorch tensors and idea needs to be extended for GroupedTensors.

- Enable te.Sequential to handle the fusion for grouped MLP block

### PR DESCRIPTION
**Pull request by vthumbe1503**
# Description

**Better Kernels for Scaled Activations**
- Scaled Srelu, Swiglu and Clamped Swiglu Pytorch Ops are replaced with their corresponding tex APIs introduced in this PR. tex APIs under the hood uses fused kernels introduced in this [PR](https://github.com/NVIDIA/TransformerEngine/pull/3132)
-  Implementaion of the tex APIs are done in activation.cpp and used directly in the corresponding scaled te ops for activations

**Precomputed Tensor Offsets Optimization**
- Tensor Offsets were already precomputed in forward pass via tex API splits_to_offset_multi. The same is done now for backward pass as well and this saves use two wasted kernel calls for calculation of offsets in the backward pass in grouped linear and 4 wasted kernel calls in grouped mlp.

Also strengthens test_grouped_mlp assertions. We expect fusion even for nvfp4_rht + scale_swiglu + glu_interleave_size = 32 case.

**BF16 Grouped MLP Performance for num_groups=8 and swiglu activation**

Metric | GPU-Bound Case (hidden_dim = 2048) | CPU-Bound Case (hidden_dim = 1024)
-- | -- | --
Tokens | 65,536 | 2,048
Before PR | 5.722 ms / iter | 1.385 ms / iter
After PR | 5.329 ms / iter | 1.047 ms / iter
Delta (%) | 🟢 -6.87% (Faster) | 🟢 -24.40% (Faster)

**MXFP8 Grouped MLP performance for num_groups = 8 and swiglu activation**

With CuteDSL fusions

Metric | GPU-Bound Case (hidden_dim = 2048) | CPU-Bound Case (hidden_dim = 128)
-- | -- | --
Tokens | 65,536 | 2,048
Before PR | 3.246 ms / iter | 1.516 ms / iter
After PR | 3.240 ms / iter | 1.461 ms / iter
Delta (%) | 🟢 -0.18% (Faster) | 🟢 -3.63% (Faster)

Without CuteDSL fusions

Metric | GPU-Bound Case (hidden_dim = 2048) | CPU-Bound Case (hidden_dim = 128)
-- | -- | --
Tokens | 65,536 | 2,048
Before PR | 7.353 ms / iter | 1.654 ms / iter
After PR | 3.398 ms / iter | 1.308 ms / iter
Delta (%) | 🟢 -53.79% (Faster) | 🟢 -20.91% (Faster)

Fixes #2988 

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Change A
- Change B

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/csrc/extensions/activation.cpp [hunk 1]
```diff
@@ -342,5 +342,149 @@ py::object clamped_dswiglu(const at::Tensor& grad, const at::Tensor& input, py::
...
+    act_func(input_nvte.data(), scales_nvte.data(), output_nvte.data(), std::forward<Args>(args)...,
+             stream);
+  });
+
+  if (quantizer.is_none()) {
+    return py::cast(output);
+  }
+  return group_quantize(output, quantizer, num_tensors, first_dims, std::nullopt, tensor_offsets,
+                        std::nullopt);
+}
```
[the remark this answers is not in the document]
Comment (vthumbe1503) [reply transformer_engine/pytorch/csrc/extensions/activation.cpp:374]: Added comment

File: transformer_engine/pytorch/ops/basic/grouped_linear.py [hunk 2]
```diff
@@ -1255,40 +1255,81 @@ def _fuser_forward_grouped_tensor(
...
+            if grouped_input.quantizer is not expected_quantizer or in_features != self.in_features:
+                raise ValueError(
+                    "GroupedLinear received an incompatible grouped input "
+                    f"(quantizer={grouped_input.quantizer}, "
+                    f"logical_shape={grouped_input.logical_shape}; "
+                    f"expected quantizer={expected_quantizer}, "
+                    f"in_features={self.in_features})"
+                )
+            grouped_x = grouped_input
+            out_shape = [total_tokens, self.out_features]
```
[the remark this answers is not in the document]
Comment (vthumbe1503) [reply transformer_engine/pytorch/ops/basic/grouped_linear.py]: Now preserving the original shape

File: transformer_engine/pytorch/ops/fused/forward_activation_grouped_linear.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+    dtype = torch.get_autocast_dtype("cuda") if torch.is_autocast_enabled() else weight.dtype
+    return linear._is_graph_safe_path_supported(
+        with_quantized_compute=True,
+        input_quantizers=input_quantizers,
+        dtype=dtype,
+        single_grouped_weight=linear.single_grouped_weight,
+    )
+
+
+class ForwardScaledActivationGroupedLinear(FusedOperation):
```
Comment (zhongbozhu) [inline transformer_engine/pytorch/ops/fused/forward_activation_grouped_linear.py]: There is a trade-off in creating a partially fused module, since we already have FC1 - ACT - FC2 fused module, We need to justify the value of creating a FC1-ACT fusion, instead of just adding features to the grouped_mlp instead.

Comment (vthumbe1503) [reply transformer_engine/pytorch/ops/fused/forward_activation_grouped_linear.py]: This essentially enables the infrastructure to enable (Act + GroupQuant from FC2) fusion, and enables to add fused kernels if possible in the future. cc: @timmoon10

Comment (vthumbe1503) [reply transformer_engine/pytorch/ops/fused/forward_activation_grouped_linear.py]: The other important thought i had in mind is the fact that ScaledActivation today doesnt take in m_splits and it is going to be a lot of upstream disruption to allow for that. And at the same time ScaledActivation can be used after Dense Layers as well and not necessarily after a GroupedLinear layer(and so it might not always need m_splits).

Allowing for this fusion, we allow the m_splits information to be also consumed in the scaled_activation + grouped quantization fusion. Right now the activation kernel that we are using isnt even using the m_splits, but for paged stashing optimization we might need that and we can potentially add a new kernel for that in the future.

File: transformer_engine/pytorch/ops/fused/backward_activation_grouped_linear.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        del basic_op_grad_extra_outputs
+        linear = self.basic_ops[0]
+        activation = self.basic_ops[1]
+        linear_ctx, activation_ctx = basic_op_ctxs
+        input_, scales = activation_ctx.saved_tensors
+        input_ = maybe_dequantize(input_, activation_ctx.dtype)
+        scales = maybe_dequantize(scales, activation_ctx.dtype)
+        grad_output = maybe_dequantize(grad_output, activation_ctx.dtype)
+
+        split_sizes = linear_ctx.saved_tensors[0]
```
[the remark this answers is not in the document]
Comment (vthumbe1503) [reply transformer_engine/pytorch/ops/fused/backward_activation_grouped_linear.py]: We early return in that case now

File: transformer_engine/pytorch/ops/fused/backward_activation_grouped_linear.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+        grouped_dy, dense_dy, grad_scales = _grouped_scaled_dactivation(
+            activation,
+            grad_output,
+            input_,
+            scales,
+            quantizer=grad_output_quantizer,
+            num_groups=linear.num_groups,
+            split_sizes=split_sizes,
+            tensor_offsets=grad_output_tensor_offsets,
+            compute_scale_grad=activation_ctx.extra_input_requires_grad,
```
[the remark this answers is not in the document]
Comment (vthumbe1503) [reply transformer_engine/pytorch/ops/fused/backward_activation_grouped_linear.py]: Forward and backward fusions are different. Forward is act + fc2; backward is fc1 + act. So linear_ctx here is fc1’s context, not fc2’s.

[the remark this answers is not in the document]
Comment (vthumbe1503) [reply transformer_engine/pytorch/ops/fused/backward_activation_grouped_linear.py]: Test is already present that verifies the correctness of grouped mlp with scaled swiglu where tensor offsets would have been different by a factor of 2 if your hypothesis is right. But it passes.

**Review by timmoon10 — APPROVED** [review]
LGTM

**Comment by vthumbe1503** [comment]
Pipeline: 61451302
